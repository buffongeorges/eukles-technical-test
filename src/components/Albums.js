import axios from "axios";
import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";

export function Albums() {
  const user = useSelector((state) => state.selectedUser);
  const [albums, setAlbums] = useState([]);
  const dispatch = useDispatch();

  useEffect(() => {
    const getUserAlbums = () => {
      if (user) {
        axios
          .get(`https://jsonplaceholder.typicode.com/users/${user.id}/albums`)
          .then((result) => {
            setAlbums(result.data);
          })
          .catch((err) => {
            console.log(err);
          });
      }
    };
    getUserAlbums();
  }, []);

  if (user) {
    return (
      <>
        <h1>Voici les albums de {user.name}</h1>
        <div style={{ display: "inline-block" }}>
          <ul>
            {albums.map((album, index) => (
              <li
                data-testid={`album-${index + 1}`}
                key={index}
                style={{ marginBottom: "1rem" }}
              >
                <Link
                  data-testid={`album-link-${index + 1}`}
                  to={"photos"}
                  onClick={() =>
                    dispatch({ type: "EDIT_SELECTED_ALBUM", newValue: album })
                  }
                >
                  {album.title}
                </Link>
              </li>
            ))}
          </ul>
        </div>
      </>
    );
  } else {
    return null;
  }
}
