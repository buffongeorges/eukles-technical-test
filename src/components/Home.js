import { useEffect, useState } from "react";
import axios from "axios";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";

export function Home() {
  const [users, setUsers] = useState([]);
  const dispatch = useDispatch();

  useEffect(() => {
    const getUsers = () => {
      axios
        .get("https://jsonplaceholder.typicode.com/users")
        .then((result) => {
          setUsers(result.data);
        })
        .catch((err) => {
          console.log(err);
        });
    };
    getUsers();
  }, []);

  return (
    <>
      <h1>La liste des utilisateurs :</h1>
      <div style={{display: 'inline-block'}}>
        <ul>
          {users.map((user, index) => (
            <li data-testid={`user-${index+1}`} key={index} style={{marginBottom: '1rem'}}><Link data-testid={`user-link-${index+1}`} to={"albums"} onClick={() => dispatch({type: "EDIT_SELECTED_USER", newValue: user})}>{user.name}</Link></li>
          ))}
        </ul>
      </div>
    </>
  );
}