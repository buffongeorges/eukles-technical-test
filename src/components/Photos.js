import axios from "axios";
import { useEffect, useState } from "react";
import { useSelector } from "react-redux";

export function Photos() {
  const user = useSelector((state) => state.selectedUser);
  const album = useSelector((state) => state.selectedAlbum);
  const [photos, setPhotos] = useState([]);
  const URL = `https://jsonplaceholder.typicode.com/albums/${album.id}/photos`;

  useEffect(() => {
    const getPhotos = () => {
      axios
        .get(URL)
        .then((result) => {
          setPhotos(result.data);
        })
        .catch((err) => {
          console.log(err);
        });
    };
    getPhotos();
  }, []);

  return (
    <>
      <h1>
        Voici les images de l'album {album.title} de l'utilisateur {user.name}:
      </h1>
      <div style={{ display: "inline-block" }}>
        <ul>
          {photos.map((album, index) => (
            <li
              data-testid={`photo-${index + 1}`}
              key={index}
              style={{ marginBottom: "1rem" }}
            >
              <img data-testid={`photo-content-${index + 1}`} src={album.url} />
            </li>
          ))}
        </ul>
      </div>
    </>
  );
}
