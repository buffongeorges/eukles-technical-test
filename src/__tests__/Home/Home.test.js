import {
  cleanup,
  render,
  screen,
  waitFor,
} from "@testing-library/react";
import { Home } from "../../components/Home";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import "@testing-library/jest-dom";
import axiosMock from "axios";
import { MemoryRouter } from "react-router-dom";

const dummyUsers = [
  {
    id: 1,
    name: "Leanne Graham",
    username: "Bret",
    email: "Sincere@april.biz",
    address: {
      street: "Kulas Light",
      suite: "Apt. 556",
      city: "Gwenborough",
      zipcode: "92998-3874",
      geo: {
        lat: "-37.3159",
        lng: "81.1496",
      },
    },
    phone: "1-770-736-8031 x56442",
    website: "hildegard.org",
    company: {
      name: "Romaguera-Crona",
      catchPhrase: "Multi-layered client-server neural-net",
      bs: "harness real-time e-markets",
    },
  },
  {
    id: 2,
    name: "Ervin Howell",
    username: "Antonette",
    email: "Shanna@melissa.tv",
    address: {
      street: "Victor Plains",
      suite: "Suite 879",
      city: "Wisokyburgh",
      zipcode: "90566-7771",
      geo: {
        lat: "-43.9509",
        lng: "-34.4618",
      },
    },
    phone: "010-692-6593 x09125",
    website: "anastasia.net",
    company: {
      name: "Deckow-Crist",
      catchPhrase: "Proactive didactic contingency",
      bs: "synergize scalable supply-chains",
    },
  },
];

afterEach(() => {
  cleanup();
});
describe(Home, () => {
  let initialState = { selectedUser: null, selectedAlbum: null };
  const mockStore = configureStore();
  let store;
  store = mockStore(initialState);

  it("should render component without crashing", () => {
    axiosMock.get.mockResolvedValue({ data: null });
    render(
      <Provider store={store}>
        <Home />
      </Provider>
    );
    const title = screen.getByRole("heading", { level: 1 });
    expect(title).toBeInTheDocument();
    expect(title).toHaveTextContent("La liste des utilisateurs :");
  });

  it("should match snapshot", () => {
    const homeComponent = (
      <Provider store={store}>
        <Home />
      </Provider>
    );
    const tree = renderer.create(homeComponent).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should display users list on api mock", async () => {
    axiosMock.get.mockResolvedValue({ data: dummyUsers });
    render(
      <Provider store={store}>
        <Home />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    const firstUserInList = await waitFor(() => screen.findByTestId("user-1"));
    expect(firstUserInList).toBeInTheDocument();
    expect(firstUserInList).toHaveTextContent("Leanne Graham");
    const secondUserInList = await waitFor(() => screen.findByTestId("user-2"));
    expect(secondUserInList).toHaveTextContent("Ervin Howell");
  });

  it("should change page on click", async () => {
    axiosMock.get.mockResolvedValue({ data: dummyUsers });
    render(
      <Provider store={store}>
        <Home />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    const firstUserInList = await waitFor(() => screen.findByTestId("user-1"));
    expect(firstUserInList).toBeInTheDocument();
    const firstUserLink = await waitFor(() =>
      screen.findByTestId("user-link-1")
    );
    expect(firstUserLink).toHaveAttribute("href", "/albums");
  });
});
