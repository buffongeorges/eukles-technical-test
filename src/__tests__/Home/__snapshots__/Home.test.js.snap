// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
Array [
  <h1>
    La liste des utilisateurs :
  </h1>,
  <div
    style={
      Object {
        "display": "inline-block",
      }
    }
  >
    <ul />
  </div>,
]
`;
